### PR TITLE
Convert `org.pkl.core.project` POJOs to `record` classes

### PR DESCRIPTION
**Pull request by KushalP**

### CODE REVIEW

File: pkl-core/src/main/java/org/pkl/core/project/CanonicalPackageUri.java [hunk 1]
```diff
@@ -71,17 +67,12 @@ public static CanonicalPackageUri of(String uriStr) throws URISyntaxException {
     return new CanonicalPackageUri(baseUri, majorVersion);
   }
 
-  public CanonicalPackageUri(URI baseUri, int majorVersion) {
-    this.baseUri = baseUri;
-    this.majorVersion = majorVersion;
-  }
-
-  @SuppressWarnings("unused")
+  @Deprecated
```
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/project/CanonicalPackageUri.java:70]: For each of these: add `forRemoval = true`, and add a doc comment about what to replace with

```suggestion
  /**
   * @deprecated As of 0.28.0, replaced by {@link #majorVersion()}.
   */
  @Deprecated(forRemoval = true)
```

File: pkl-core/src/main/java/org/pkl/core/project/Package.java [hunk 2]
```diff
@@ -25,113 +25,93 @@
...
-    this.sourceCode = sourceCode;
-    this.sourceCodeUrlScheme = sourceCodeUrlScheme;
-    this.license = license;
-    this.licenseText = licenseText;
-    this.issueTracker = issueTracker;
-    this.apiTests = apiTests;
-    this.exclude = exclude;
-  }
-
+public record Package(
```
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/project/Package.java:28]: We can remove `equals`, `hashCode`, and `toString` for this one right?

**Review by bioball — CHANGES_REQUESTED** [review]
There's a bit more polish here to be done here, but overall looks good!

File: pkl-core/src/main/java/org/pkl/core/project/Package.java (hunk 2 above)
Comment (KushalP) [reply pkl-core/src/main/java/org/pkl/core/project/Package.java:28]: I've removed `equals` and `hashCode`. The `toString` function on `record` uses square brackets `MyObjectName[...]` whereas this one uses curly braces `Package{...}`. What would you prefer to do here?

Comment (bioball) [reply pkl-core/src/main/java/org/pkl/core/project/Package.java:28]: It's fine to just use the one on `record`; the `toString()` is only meant for debugging purposes anyway!

File: pkl-core/src/main/java/org/pkl/core/project/CanonicalPackageUri.java [hunk 3]
```diff
@@ -71,17 +67,15 @@ public static CanonicalPackageUri of(String uriStr) throws URISyntaxException {
...
+  /**
+   * @deprecated As of 0.28.0, replaced by {@link #majorVersion()}.
+   */
+  @Deprecated(forRemoval = true)
   public int getMajorVersion() {
     return majorVersion;
   }
 
-  @SuppressWarnings("unused")
+  @Deprecated
```
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/project/CanonicalPackageUri.java:78]: Add `forRemoval = true, add doc comment

File: pkl-core/src/main/java/org/pkl/core/project/DeclaredDependencies.java [hunk 4]
```diff
@@ -21,35 +21,27 @@
...
   public Map<String, RemoteDependency> getRemoteDependencies() {
     return remoteDependencies;
   }
 
+  @Deprecated
   public URI getProjectFileUri() {
     return projectFileUri;
   }
 
+  @Deprecated
```
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/project/DeclaredDependencies.java:44]: Add `forRemoval = true, add doc comment

File: pkl-core/src/main/java/org/pkl/core/project/Package.java [hunk 5]
```diff
@@ -18,209 +18,100 @@
...
   public @Nullable URI getIssueTracker() {
     return issueTracker;
   }
 
+  @Deprecated
   public List<Path> getApiTests() {
     return apiTests;
   }
 
+  @Deprecated
```
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/project/Package.java:113]: Add `forRemoval = true, add doc comment

File: pkl-core/src/main/java/org/pkl/core/project/ProjectPackager.java [hunk 6]
```diff
@@ -452,42 +451,33 @@ public void validateImportsAndReads(Project project, Path pklModulePath) {
...
     public Path getMetadataFile() {
       return metadataFile;
     }
 
+    @Deprecated
     public Path getMetadataChecksumFile() {
       return metadataChecksumFile;
     }
 
+    @Deprecated
```
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/project/ProjectPackager.java:480]: Add `forRemoval = true, add doc comment

**Review by bioball — CHANGES_REQUESTED** [review]
Some cleanup work needed!

**Review by bioball — APPROVED** [review]
LGTM!

**Pull request merged**